[GILJOB-1] 프론트엔드 개발 환경 구성

raejoonee opened the pull request:
## Summary
프로젝트 프론트엔드 개발 환경 구성을 완료했습니다 🎉 
dev build, production build가 정상적으로 완료되는 것을 확인했습니다 😎 

## Detail
### npm d9aa6f5
* `npm init` 명령어를 사용해서 기본 환경 설정을 완료했습니다.
### GitHub c04ea4e
* PR 템플릿을 작성했습니다. `develop` 브랜치로 merge가 완료되면, develop으로 merge되는 모든 PR의 설명란에 자동으로 템플릿이 먼저 작성됩니다.
* `.gitignore`를 작성했습니다. bafae0f
### VSCode 147977a
* 각자 사용하고 있는 VSCode의 `settings.json`의 내용에 따라 Lint가 적용되었다가 풀렸다가 할 수 있으므로 통일된 규칙을 따르도록 했습니다.
### babel d63935a
* babel 환경의 경우, [카카오 테크 블로그의 좋은 글](https://tech.kakao.com/2020/12/01/frontend-growth-02/)을 참고하여 전역 오염 없는 폴리필을 지원하도록 corejs3으로 설정했습니다.
### Webpack a1540e2
* Webpack 기본 환경을 설정할 때, dev build와 production build에서 공통으로 설정되는 부분은 `config/webpack/webpack.base.js`에 작성했습니다.
* [Webpack 공식 문서의 style-loader 항목](https://webpack.js.org/loaders/style-loader/#recommend)에 따르면, dev build에서만 style-loader를 사용하는 것을 권장하고 있습니다. 따라서 production build에선 MiniCssExtractPlugin를 사용하도록 설정했습니다.
### SCSS bd0a744
### TypeScript 08bd9e0
* `tsconfig.json`을 작성할 때,  미리 `@types` 외부 모듈 폴더 사용에 대비할 수 있도록 했습니다.
* production build 시, TypeScript 코드를 컴파일하도록 `tsc`를 추가했습니다.
### Prettier c51a2c2
### ESLint 3fee24c 8a655be
### React 93fe4e5
* 정상적으로 프로젝트 환경 설정이 마무리됐는지 확인하기 위해 `Hello world!`를 출력하는 테스트 컴포넌트를 작성했습니다. d8ad4f6

## Todo
Storybook, Jest, GitHub Actions(CI) 환경 설정은 내일 하도록 하겠습니다 😢 

리뷰 꼭 부탁드리고, 궁금한 점이 있거나 코드 개선점이 있다면 언제든지 댓글 마구마구 달아주세요 👨‍💻

raejoonee committed:
[GILJOB-14] env: 초기 npm 기본 설정을 완료한다.

raejoonee committed:
[GILJOB-17] env: PR 템플릿을 작성한다.

raejoonee committed:
[GILJOB-17] env: .gitignore를 작성한다.

raejoonee committed:
[GILJOB-25] env: VSCode에서 파일을 저장할 때 개인 설정보다 ESLint 설정을 우선하여 따르도록 한다.

raejoonee committed:
[GILJOB-18] env: babel 기본 환경을 설정한다.

raejoonee committed:
[GILJOB-18] env: Webpack 기본 환경을 설정한다.

dev build에서 style-loader를 사용하고, production build에서 MiniCssExtractPlugin를 사용한다.

raejoonee committed:
[GILJOB-19] env: SCSS 기본 환경을 설정한다.

raejoonee committed:
[GILJOB-20] env: TypeScript 기본 환경을 설정한다.

raejoonee committed:
[GILJOB-22] env: Prettier 기본 환경을 설정한다.

raejoonee committed:
[GILJOB-22] env: ESLint 기본 환경을 설정한다.

raejoonee committed:
[GILJOB-22] env: ESLint 예외 경로를 설정한다.

raejoonee committed:
[GILJOB-26] env: React 기본 환경을 설정한다.

raejoonee committed:
[GILJOB-26] env: React 환경 설정이 정상적으로 완료되었는지 확인하기 위한 테스트 컴포넌트를 작성한다.

kyoung-jnn commented on package.json:
```diff
@@ -0,0 +1,60 @@
+    "@babel/core": "^7.15.5",
+    "@babel/plugin-transform-runtime": "^7.15.0",
+    "@babel/preset-env": "^7.15.6",
+    "@babel/preset-flow": "^7.14.5",
```
저희 typescript사용하니까 flow uninstall 해도 괜찮지 않나요?
`@babel/preset-typescript` + `@babel/preset-flow` 이렇게 사용해도 괜찮나여?

raejoonee replied:
이렇게 사용해도 괜찮긴 한데, 지적해주신 대로 uninstall하는 것이 좋아 보이네요!

kyoung-jnn commented on config/webpack/webpack.dev.js:
```diff
@@ -0,0 +1,45 @@
+      },
+    ],
+  },
+  plugins: [
+    new HtmlWebPackPlugin({
+      template: path.resolve(TEMPLATE_PATH, 'index.html'),
+      minify: {
+        removeComments: true,
+      },
+    }),
+  ],
+});
```
`new webpack.ProvidePlugin({
      React: 'react'
    })` 
ProvidePlugin 플러그인 사용하면 `import React from 'react'` 안해도 되요!

raejoonee replied:
헉 너무 좋은 지적 감사합니다 😊 바로 수정하도록 할게요!

raejoonee replied:
jest 때문에 `import React from 'react'`를 필수로 써야 할 것 같습니다 😢 
CRA를 쓰지 않은 React 프로젝트에 대하여 `import React from 'react'` 없이 테스트 환경 설정을 구축하기 힘드네요 😭

kyoung-jnn commented on config/webpack/webpack.base.js:
```diff
@@ -0,0 +1,33 @@
+    clean: true,
+  },
+  module: {
+    rules: [
+      {
+        test: /\.(ts|js)x?$/,
+        include: path.resolve('src'),
+
+        exclude: /node_modules/,
+        use: {
+          loader: 'babel-loader',
+        },
+      },
+    ],
```
`ts-loader` 없이도 가능한지 처음 알았네요!

https://github-wiki-see.page/m/woowa-techcamp-2021/store-4/wiki/babel-loader-%26-%40babel-preset-typescript-vs-ts-loader-%EA%B0%84%EB%8B%A8%ED%95%9C-%EC%A0%95%EB%A6%AC

kyoung-jnn commented on tsconfig.json:
```diff
@@ -0,0 +1,21 @@
+    "strict": true,
+    "rootDir": "./",
+    "outDir": "./dist",
+    "baseUrl": "./",
```
baseUrl 을 **`./src`** 로 하는게 편하지 않을까요?

raejoonee replied:
`"baseUrl": "./"`은 타입스크립트 코드가 모듈을 import할 때 `tsconfig.json`의 위치에서 시작하는 상대 경로를 쓰도록 하는 효과가 있어서 이렇게 두었습니다. 이 덕분에 `../`, `./`를 덜 사용하게 됩니다.

[이 링크](https://www.typescriptlang.org/tsconfig#baseUrl)를 참조하시면 좋을 것 같습니다.

raejoonee replied:
일단은 이대로 두겠습니다. Webpack이 모듈의 경로를 탐색할 때 alias를 사용하고 있어서 굳이 바꿀 필요는 없어 보입니다.

raejoonee committed:
[GILJOB-18] fix: @babel/preset-flow를 사용하지 않도록 수정한다.

https://github.com/YAPP-19th/Web-Team-1-Frontend/pull/1#discussion_r717348551

raejoonee committed:
[GILJOB-18] fix: ProvidePlugin을 사용해 React를 import하지 않아도 되도록 수정한다.

https://github.com/YAPP-19th/Web-Team-1-Frontend/pull/1#discussion_r717366227
추가로, 기존에 있던 테스트 코드에서도 import React를 제거했습니다.

raejoonee committed:
[GILJOB-20] fix: tsconfig 환경 설정을 수정한다.

* commonjs 대신 esnext를 사용하도록 수정
* import React를 하지 않기 때문에 jsx를 "react-jsx"로 수정
* 그 밖에 지켜야 할 규칙들 추가

raejoonee committed:
[GILJOB-18] fix: Webpack 설정 파일에 있는 import 오류를 수정한다.

raejoonee committed:
[GILJOB-21] env: GitHub Actions를 사용하여 CI pipeline의 일부를 구성한다.

raejoonee committed:
[GILJOB-27] env: Jest를 사용하기 위해 의존 모듈을 설치하고 Jest 개발 환경을 구성한다.

raejoonee committed:
[GILJOB-27] test: 테스트 컴포넌트가 정상적으로 렌더링되는지 테스트한다.

raejoonee committed:
[GILJOB-27] fix: Jest 환경을 위해서 React를 다시 import하도록 수정한다.

raejoonee committed:
[GILJOB-21] env: CI 과정에 테스트를 추가한다.

LeeJuhae commented on config/webpack/webpack.base.js:
```diff
@@ -0,0 +1,33 @@
+  },
+  resolve: {
+    extensions: ['.js', '.css', '.scss', '.ts', '.jsx', '.tsx'],
+    alias: {
```
#1 (comment)
말씀하신 alias 사용하는 부분이 여기 맞나요?
package.json의 scripts에서 webpack.dev.js, webpack.prod.js만 사용되는 것 같은데
webpack.base.js는 사용되지는 않는건가요?

LeeJuhae reviewed (approved):
제가 아직 모르는 것이 많아 모두에게 도움이 되는 리뷰를 쓰지 못해서 아쉽습니다...
앞으로 꾸준히 공부하고, 프로젝트 진행하며 @raejoonee 님이 셋팅해주신 개발 환경 관련해서 궁금한 점 있으면 추후에라도 여쭤보겠습니다!
작업 관련 summary까지 명확하고 깔끔하게 작성해주셔서 리뷰하며 webpack, babel 등 많은 공부할 수 있었습니다. 🙏🏻

kyoung-jnn commented:
이미지 파일 처리를 위한 `file-loader, url-loader` 도 필요해 보입니다!

raejoonee committed:
fix: jest 설정을 수정한다.

https://github.com/YAPP-19th/Web-Team-1-Frontend/pull/2#discussion_r721139568
https://stackoverflow.com/questions/65947682/expect-tohaveattribute-is-not-a-function-why

raejoonee committed:
fix: App.tsx에서 React를 import하지 않도록 수정한다.

kyoung-jnn committed:
[GILJOB-18] env: Webpack 기본 환경 설정

file-loader, url-loader 추가

raejoonee committed:
fix: App.tsx에서 React를 import하도록 수정한다.

raejoonee committed:
Merge branch 'env/base' into env/ci

raejoonee committed:
Fix: package-lock.json 파일을 수정한다.

raejoonee committed:
Merge pull request #2 from YAPP-19th/env/ci

[GILJOB-1] 프로젝트 CI 구성 및 테스트 환경 설정
